cpp/src/merge/merge.cu

```diff
@@ -13,417 +13,509 @@
+ 
+     //extract merged row order according to indices:
+     //
+     const auto merged_indices = generate_merged_indices(index_left_view,
+                                                         index_right_view,
+                                                         column_order,
+                                                         null_precedence,
+                                                         nullable);
+     //create merged table:
+     //
+     std::vector<std::unique_ptr<column>> merged_cols;
+     merged_cols.reserve(n_cols);
+ 
+     const column_merger merger{merged_indices, mr, stream};
+     transform(left_table.begin(), left_table.end(), 
+              right_table.begin(), 
+              std::back_inserter(merged_cols),
+             [&](auto& left_col, auto& right_col){
+               return cudf::experimental::type_dispatcher(left_col.type(),
                                                           merger,
                                                           left_col,
                                                           right_col);
-        v_merged_cols.emplace_back(std::move(merged));
-      }
-
-    return std::make_unique<cudf::experimental::table>(std::move(v_merged_cols));
-}
-
-}  // namespace detail
-
-std::unique_ptr<cudf::experimental::table> merge(table_view const& left_table,
-                                                 table_view const& right_table,
-                                                 std::vector<cudf::size_type> const& key_cols,
-                                                 std::vector<cudf::order> const& column_order,
-                                                 std::vector<cudf::null_order> const& null_precedence,
-                                                 rmm::mr::device_memory_resource* mr){
-  return detail::merge(left_table, right_table, key_cols, column_order, null_precedence, mr);
-}
-
-}  // namespace experimental
-}  // namespace cudf
+             });
+ 
+     return std::make_unique<cudf::experimental::table>(std::move(merged_cols));
+ }
+ 
+ class merge_queue_item {
+ private:
+   table_view view_;
+   table_ptr_type table_;
+   cudf::size_type priority_;
+ 
+   void set_priority() {
+     // Smallest tables have the highest priority
+     priority_ = -view_.num_rows();
+   }
+ 
+ public:
+   merge_queue_item(table_view const& view, table_ptr_type&& table):
+   view_(view), 
+   table_(std::move(table)) {
+     set_priority();
+   }
+
+   explicit merge_queue_item(std::priority_queue<merge_queue_item>& queue){
+     auto& item = const_cast<merge_queue_item&>(queue.top());
+     view_ = item.view_;
+     // Safe as priority_ is not affected by the move
+     table_ = std::move(item.table_);
+     set_priority();
+
+     queue.pop();
+   }
+
+   auto& view() const { return view_; }
+   auto& table() { return table_; }
+ 
+   bool operator<(merge_queue_item const& other) const {
+       return priority_ < other.priority_;
```

Comment:
Is the extra `priority_` member necessary? Can't this function just do:
```
   return this->view_.num_rows() > other.view_.num_rows()
```

Reply:
The reason why moving out of the priority queue does not exist is because the state of the object moved from might not keep the property that puts it in the given place in the queue. So I add this to make sure the priority does not change because of the move.